### ISSUE TITLE
Files are considered to be renamed when using rev1:file..rev2:file

### ISSUE THREAD

**Issue by blueyed** [so-fancy/diff-so-fancy#149]
``` diff
% git --no-pager diff stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin/ptw-watch-tests
diff --git a/stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin/ptw-watch-tests b/update-ptw-watch-tests:bin/ptw-watch-tests
index f59eee5..4435331 100644
--- a/stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin/ptw-watch-tests
+++ b/update-ptw-watch-tests:bin/ptw-watch-tests
…
```

With g-d-f the header is displayed as:

```
 {stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin => update-ptw-watch-tests:bin}/ptw-watch-tests | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

───────────── …
renamed: stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin/ptw-watch-tests to update-ptw-watch-tests:bin/ptw-watch-tests
───────────── …
```

I think any `^\w:` should be removed when comparing files here.

**Comment by scottchiefbaker** [comment]
The issue here appears to be the `stash:` on the filename? I've never done a diff on a stash before. I'm not even sure how to read this syntax. The line in question is:

    diff --git a/stash:bin/ptw-watch-tests..update-ptw-watch-tests:bin/ptw-watch-tests b/update-ptw-watch-tests:bin/ptw-watch-tests

Currently the logic for detecting the file names is [here](https://github.com/so-fancy/diff-so-fancy/blob/master/lib/diff-so-fancy.pl#L55). Maybe you can help translate what this *should* look like. That is a gnarly filename/repo combination.

**Comment by Blaisorblade** [comment]
FWIW: I `stash` is just a (special) reference, but arbitrary "treeish" objects could be there — we only know that colon is used as a separator. Unfortunately, on Linux `:` in filenames *is* allowed, though it's a very bad idea (among other things, it doesn't work on OS X, nor I think on Windows).

Worse, `diff-so-fancy` seems to dislike `:` anyway:
```
$ git add a:b
$ git diff --staged
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
$ git --no-pager diff --staged
diff --git a/a:b b/a:b
new file mode 100644
index 0000000..e69de29
$ git rm -f a:b
rm 'a:b'
```

Here's an example of a diff with stranger treeish (though this is from an actual rename, so not very helpful):

```
$ git --no-pager diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
diff --git a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
index 6c34f02..6e78113 100644
--- a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
+++ b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
@@ -1,8 +1,8 @@
...
```

**Comment by OJFord** [comment]
Colons in filenames is a rabbithole not worth anyone going out of their way to support, IMO. Just recently there was a blog post about a git/NTFS issue whereby cloning such a repo on Windows caused all manner of problems, since `:` is the syntax for an alternate data stream; so an 'empty' file was created with the contents existing but semi-hidden.

If we assume colons appear only to separate refs from filenames, then a rename is anything that matches `a/<ref1>:<fname1> b/<ref2>:<fname2`; while `a/<ref1>:<fname> b/<ref2>:<fname>` is _not_ a renaming.

**Comment by Blaisorblade** [comment]
> Just recently there was a blog post about a git/NTFS issue whereby cloning such a repo on Windows caused all manner of problems, since : is the syntax for an alternate data stream; so an 'empty' file was created with the contents existing but semi-hidden.

Ah right! [Link](http://latkin.org/blog/2016/07/20/git-for-windows-accidentally-creates-ntfs-alternate-data-streams/). I agree on ignoring this circumstance, even though I raised the point.

> If we assume colons appear only to separate refs from filenames, then a rename is anything that matches a/<ref1>:<fname1> b/<ref2>:<fname2>; while a/<ref1>:<fname> b/<ref2>:<fname> is not a renaming.

That makes sense, though my above example (and even the one in the issue) doesn't quite match if I get you correctly—it matches instead `diff -git a/<ref1>:<fname1>..<ref2>:<fname2> b/<ref2>:<fname2>` (which maybe is a bug; observed with git 2.9.0 on a checkout of https://github.com/agda/agda).
```
$ git --no-pager diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
diff --git a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
```

**Comment by OJFord** [comment]
> though my above example (and even the one in the issue) doesn't quite match if I get you correctly—it matches instead `diff -git a/<ref1>:<fname1>..<ref2>:<fname2> b/<ref2>:<fname2>`

Oh yes, sorry, I missed that. I'm having trouble parsing that, not sure what it means at all; it's as if one side of the diff is a diff in its own right.

Does this issue still occur without that `..` complexity, just between refs?

**Comment by Blaisorblade** [comment]
> I'm having trouble parsing that, not sure what it means at all; it's as if one side of the diff is a diff in its own right.

I don't get it either and currently suspect it might not be intended.

> Does this issue still occur without that .. complexity, just between refs?

Good question! Replacing `..` by a space removes this second issue with `git diff` and only leaves the diff-so-fancy issue.
Wild conjecture: the first filename is taken by taking every non-colon from the argument before splitting, that is, `[^:]*:`, but the `*` is overly greedy.

Without `..`:
```
$ git diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt 777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
──── …
renamed: 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt to 777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
──── …
$ git --no-pager diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt 777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
diff --git a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
index 6c34f02..6e78113 100644
--- a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt
+++ b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
```

With `..`:

```
$ git diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
──── …
renamed: 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt to 777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
──── …
$ git --no-pager diff 777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
diff --git a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
index 6c34f02..6e78113 100644
--- a/777c04865194f345c94b94f9a3e8bf627da11480~:doc/release-notes/2-5-4.txt..777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
+++ b/777c04865194f345c94b94f9a3e8bf627da11480:doc/release-notes/2-6-1.txt
```

**Comment by Blaisorblade** [comment]
Regarding my previous conjecture on the bug: of course I have to retract it, the parsing code is much more insane than reasonable. Not 100% sure what happens, but there is a pointer to the argument used as the first part, and at some points the argument is modified to be split in the middle, and at some points it's unsplit. One must love NULL-termined strings.

https://github.com/git/git/blob/f8f7adce9fc50a11a764d57815602dcb818d1816/revision.c#L1447
https://github.com/git/git/blob/f8f7adce9fc50a11a764d57815602dcb818d1816/revision.c#L1451
https://github.com/git/git/blob/f8f7adce9fc50a11a764d57815602dcb818d1816/revision.c#L1480
https://github.com/git/git/blob/f8f7adce9fc50a11a764d57815602dcb818d1816/revision.c#L1524

**Comment by scottchiefbaker** [comment]
Can we re-test this issue on `next`? I updated the file rename code in that branch.

**Comment by scottchiefbaker** [comment]
ping @blueyed

**Comment by blueyed** [comment]
@scottchiefbaker 
Should I test this with `next` or master (by now)?

**Comment by scottchiefbaker** [comment]
Always `next`

**Comment by scottchiefbaker** [comment]
I believe this is fixed in v1.2.5 so I'm closing this issue. If not please feel free to re-open this issue.
